sycl/doc/extensions/experimental/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -0,0 +1,634 @@
+
+The aim of this document is to propose a mechanism for users to request the
+fusion of two or more kernels in a SYCL graph into a single kernel **at
+runtime**. This requires the extension of the runtime with some sort of JIT
+compiler to allow for the fusion of kernel functions at runtime.
+
+== 6. Specification
+
+=== 6.1. Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GRAPH_FUSION` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== 6.2. API modifications
+
+==== 6.2.1. Properties
+
+===== 6.2.1.1. Graph Fusion Property
+
+The API for `command_graph<graph_state::modifiable>::finalize()` includes a
+`property_list` parameter. The following property, defined by this extension,
+can be added to the property list to indicate that the kernels in the
+command-graph should be fused. 
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::perform_fusion 
+```
+
+The property is not prescriptive. Implementations are free to not perform fusion
+if it is not possible (see below section <<_6_5_limitations>>), fusion is not
+supported by the implementation, or the implementation decides not to perform
+fusion for other reasons. It is not an error if an implementation does not
+perform fusion even though the property is passed. 
+
+Implementations can provide a diagnostic message in case fusion was not
+performed through an implementation-specified mechanism, but are not required to
+do so.
+
+===== 6.2.1.2. Barrier property
+
+The following property can be added to the `property_list` of the
+`command_graph<graph_state::modifiable>::finalize()` API.
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::no_barriers
+```
+
+If the property list contains this property, no barriers are introduced between
+kernels in the fused kernel (see below section on synchronization on kernels). 
+
+The property only takes effect if the
+`sycl::ext::oneapi::experimental::property::command_graph::perform_fusion`
+property is also part of the `property_list` of the same invocation of
+`command_graph<...>::finalize()`. 
+
+===== 6.2.1.3. Local internalization property
+
+The following property can be passed to three different APIs, namely:
+
+* The `accessor` constructor, giving a more granular control
+* The `buffer` constructor, in which case all the `accessors` derived from 
+this buffer will inherit this property (unless overridden).
+* The `property_list` parameter of 
+`command_graph<graph_state::modifiable>::add_malloc_device()` to apply the 
```

Comment:
We are considering removing this `add_malloc_device` API (due to issue outlined [here](https://github.com/intel/llvm/pull/5626#discussion_r1084170330)), and instead advising users to use core `sycl::malloc_device` outside the graph. i.e tending towards option 0 from the call.

Would using the `property_list` parameter to `sycl::malloc_device`  to pass the internalization  information, and then using that device pointer in a graph be satisfactory for fusion?

Reply:
As long as it is possible for the implementation to determine that a pointer was created with an internalization property, using `sycl::malloc_device` to allocate USM pointers for internalization should be fine for fusion. 

The downside of using `sycl::malloc_device` with the internalization properties outside of the graph is mainly granularity: When passing the properties to `add_malloc_device`, the additional semantics of the properties are limited to one graph. If, on the other hand, the property is passed `sycl::malloc_device`, it will apply to *all* graphs that use the pointer.

Reply:
I agree the granularity isn't ideal, but good to know that the finer granularity is a nice-to-have rather must-have for you. Right now I don't think we have confidence in any of our potential graph memory allocation designs, and am reluctant to specify something that user applications start using, and we then have to depreciate (even through the extension is experimental). So if core SYCL USM is good enough, it may be the best way forward for now so we that can focus on the critical path to get something into the hands of users. Then revisit this again once we have real world feedback.

Reply:
Yes, core SYCL USM should be fine as long as there is a way to match a USM pointer with its properties specified during allocation. I'm no expert on the implementation of USM, so I can't predict how complicated that matching is. If device allocation from multiple devices are in play, the same address could actually be used for two completely different pointers from two different devices, making the matching more complicated. But this is probably an implementation detail. 

Regarding granularity, a while back we had also considered properties for USM internalization, which would simply take a pointer as value of the property, and that you could pass to `complete_fusion` (or `finalize` for this proposal). 
Something along the lines of:
`sycl::ext::oneapi::experimental::property::promote_private(void* ptr)`. 

This would allows to limit the internalization semantic for a specific pointer to just this graph. However, IMHO it is less elegant than a data-less property and would also mean you need two separate properties for USM and buffer internalization (unless you allow `nullptr` to be passed).

Reply:
Another advantage of having `add_malloc_device` and passing the properties to that API: You can completely eliminate the allocation of the device pointer if internalization is successful. After the JIT compilation as part of `finalize`, it is known whether the allocation is needed at all. If it is not needed, it can be removed from the graph, saving some time for the device allocation and improving performance. 

This optimization is not possible if the USM allocations are not part of the graph and happen outside of the graph with the core SYCL USM API.

Reply:
We have a draft PR for this change now, feel free to comment - https://github.com/reble/llvm/pull/79 Can hold off on merging this until we have our sync-up call if you'd like to discuss it F2F

Reply:
Thanks! I've put a comment on https://github.com/reble/llvm/pull/79, might be worth to discuss this further in the sync-up meeting.